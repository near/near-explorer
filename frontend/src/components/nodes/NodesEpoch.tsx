import * as React from "react";
import { useDateFormat } from "../../hooks/use-date-format";

import { Row, Col } from "react-bootstrap";

import ProgressBar from "../utils/ProgressBar";

import { useTranslation } from "react-i18next";
import { styled } from "../../libraries/styles";
import { useSubscription } from "../../hooks/use-subscription";
import { parseISO } from "date-fns";

const NodesEpochContent = styled(Col, {
  margin: "15px 0",
});

const TextValue = styled("span", {
  color: "#37dbf4",
});
const TextValueRemainedPercent = styled(TextValue, {
  fontWeight: 700,
});

const NodesEpochRow = styled(Row, {
  backgroundColor: "#292526",
  color: "#d5d4d8",
  fontSize: 16,
  fontWeight: 500,
});

const NodesEpochLineProgress = styled(ProgressBar, {
  paddingHorizontal: 0,
  height: 4,
});

const NodesEpochCircleProgress = styled(ProgressBar, {
  marginLeft: "auto",
  width: 40,
});

const NodesEpochCircleProgressLabel = styled("span", {
  color: "#d5d4d8",
});

interface Props {
  epochLength: number;
  epochStartHeight: number;
  epochStartTimestamp: number;
}

const NodesEpoch: React.FC<Props> = React.memo(
  ({ epochStartHeight, epochLength, epochStartTimestamp }) => {
    const { t } = useTranslation();
    const latestBlockSub = useSubscription(["latestBlock"]);
    const format = useDateFormat();
    if (latestBlockSub.status !== "success") {
      return null;
    }
    const epochProgress =
      ((latestBlockSub.data.height - epochStartHeight) / epochLength) * 100;
    const timeRemaining =
      ((latestBlockSub.data.timestamp - epochStartTimestamp) / epochProgress) *
      (100 - epochProgress);

    return (
      <NodesEpochRow>
        <NodesEpochContent xs="12">
          <Row noGutters>
            <Col xs="7">
              <Row className="d-none d-md-flex">
                <Col>
                  {t("component.nodes.NodesEpoch.current_epoch_start") + ": "}
                  <TextValue>
                    {t("component.nodes.NodesEpoch.block") + " #"}
                    {epochStartHeight}
                  </TextValue>
                </Col>
              </Row>

              <Row className="d-xs-flex d-md-none">
                <Col xs="12">
                  {t("component.nodes.NodesEpoch.current_epoch_start")}
                </Col>
                <Col xs="12">
                  <TextValue>
                    {t("component.nodes.NodesEpoch.block") + " #"}
                    {epochStartHeight}
                  </TextValue>
                </Col>
              </Row>
            </Col>

            <Col sm="5" className="text-right d-none d-md-block ">
              <TextValueRemainedPercent>
                {epochProgress.toFixed(0)}
                {"% " + t("component.nodes.NodesEpoch.complete")}
              </TextValueRemainedPercent>
              {` (${format(timeRemaining, "HH:mm:ss", { utc: true })} ${t(
                "component.nodes.NodesEpoch.remaining"
              )})`}
            </Col>

            <Col xs="5" className="text-right d-xs-block d-md-none">
              <NodesEpochCircleProgress
                percent={epochProgress}
                strokeColor="#37dbf4"
                trailColor="transparent"
                type="circle"
                strokeWidth={4}
                label={
                  <NodesEpochCircleProgressLabel>
                    {epochProgress.toFixed(0)}%
                  </NodesEpochCircleProgressLabel>
                }
              />
            </Col>
          </Row>
        </NodesEpochContent>
        <Col xs="12" className="d-none d-md-block px-0">
          <NodesEpochLineProgress
            percent={epochProgress}
            strokeColor="#37dbf4"
            trailColor="transparent"
          />
        </Col>
      </NodesEpochRow>
    );
  }
);

export default NodesEpoch;
