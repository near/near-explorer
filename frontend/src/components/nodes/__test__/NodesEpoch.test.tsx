import renderer from "react-test-renderer";

import NodesEpoch from "../NodesEpoch";

describe("<NodesEpoch />", () => {
  it("renders", () => {
    expect(
      renderer.create(
        <NodesEpoch
          epochStartBlock={{
            gasUsed: "5728980156616",
            gasPrice: "100000000",
            hash: "Hk7qfJqqn1BV2hsz3a4s5a7tmcCgSDx4PVf3T6mcMTf6",
            height: 36647454,
            prevHash: "EGKpS8AguX6DWTGcX2LAsv2oJTggJef49ojjStRqMQsQ",
            timestamp: 1620305916060,
            totalSupply: "1026697138447885703199686999754199",
            transactionsCount: 1,
          }}
          epochStartHeight={36647454}
        />
      )
    ).toMatchSnapshot();
  });
});
