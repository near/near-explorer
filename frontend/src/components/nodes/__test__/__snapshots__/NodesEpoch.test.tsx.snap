// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodesEpoch /> renders 1`] = `
<div
  className="nodes-epoch row"
>
  <div
    className="nodes-epoch-content col-12"
  >
    <div
      className="row no-gutters"
    >
      <div
        className="col-7"
      >
        <div
          className="d-none d-md-flex row"
        >
          <div
            className="col"
          >
            Current Epoch Start: 
            <span
              className="text-value"
            >
              Block #
              36647454
            </span>
          </div>
        </div>
        <div
          className="d-xs-flex d-md-none row"
        >
          <div
            className="col-12"
          >
            Current Epoch Start
          </div>
          <div
            className="col-12"
          >
            <span
              className="text-value"
            >
              Block #
              36647454
            </span>
          </div>
        </div>
      </div>
      <div
        className="text-right d-none d-md-block  col-sm-5"
      >
        <span
          className="text-value persnt-remains"
        >
          50
          % complete
        </span>
         (06:00:00 remaining)
      </div>
      <div
        className="text-right d-xs-block d-md-none col-5"
      >
        <div
          className="progress-bar circle node-epoch-circle-progress"
        >
          <svg
            className="rc-progress-circle progress-bar-content"
            label={
              <span
                className="circle-progress-label"
              >
                50
                %
              </span>
            }
            style={Object {}}
            type="circle"
            viewBox="0 0 100 100"
          >
            <path
              className="rc-progress-circle-trail"
              d="M 50,50 m 0,-48
   a 48,48 0 1 1 0,96
   a 48,48 0 1 1 0,-96"
              fillOpacity="0"
              stroke="transparent"
              strokeLinecap="round"
              strokeWidth={1}
              style={
                Object {
                  "stroke": "transparent",
                  "strokeDasharray": "301.59289474462014px 301.59289474462014px",
                  "strokeDashoffset": "-0px",
                  "transition": "stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s",
                }
              }
            />
            <path
              className="rc-progress-circle-path"
              d="M 50,50 m 0,-48
   a 48,48 0 1 1 0,96
   a 48,48 0 1 1 0,-96"
              fillOpacity="0"
              opacity={1}
              stroke=""
              strokeLinecap="round"
              strokeWidth={4}
              style={
                Object {
                  "stroke": "#37dbf4",
                  "strokeDasharray": "150.79644737231007px 301.59289474462014px",
                  "strokeDashoffset": "-0px",
                  "transition": "stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s",
                }
              }
            />
          </svg>
          <div
            className="progress-bar-label"
          >
            <span
              className="circle-progress-label"
            >
              50
              %
            </span>
          </div>
        </div>
        <style
          global={true}
          jsx={true}
        >
          
          .progress-bar {
            background: transparent;
            position: relative;
          }

          .progress-bar .progress-bar-content {
            background: transparent;
          }

          .progress-bar .rc-progress-line-path,
          .progress-bar .rc-progress-circle-path {
            stroke-linecap: square;
          }

          .progress-bar .progress-bar-label {
            position: absolute;
            width: 100%;
            margin: auto;
            font-size: 12px;
            font-weight: 500;
          }
        
        </style>
      </div>
    </div>
  </div>
  <div
    className="d-none d-md-block px-0 col-12"
  >
    <div
      className="progress-bar line node-epoch-line-progress"
    >
      <svg
        className="rc-progress-line progress-bar-content"
        preserveAspectRatio="none"
        style={Object {}}
        viewBox="0 0 100 1"
      >
        <path
          className="rc-progress-line-trail"
          d="M 0.5,0.5
         L 99.5,0.5"
          fillOpacity="0"
          stroke="transparent"
          strokeLinecap="round"
          strokeWidth={1}
        />
        <path
          className="rc-progress-line-path"
          d="M 0.5,0.5
         L 99.5,0.5"
          fillOpacity="0"
          stroke="#37dbf4"
          strokeLinecap="round"
          strokeWidth={1}
          style={
            Object {
              "strokeDasharray": "49.5px, 100px",
              "strokeDashoffset": "-0px",
              "transition": "stroke-dashoffset 0.3s ease 0s, stroke-dasharray .3s ease 0s, stroke 0.3s linear",
            }
          }
        />
      </svg>
    </div>
    <style
      global={true}
      jsx={true}
    >
      
          .progress-bar {
            background: transparent;
            position: relative;
          }

          .progress-bar .progress-bar-content {
            background: transparent;
          }

          .progress-bar .rc-progress-line-path,
          .progress-bar .rc-progress-circle-path {
            stroke-linecap: square;
          }

          .progress-bar .progress-bar-label {
            position: absolute;
            width: 100%;
            margin: auto;
            font-size: 12px;
            font-weight: 500;
          }
        
    </style>
  </div>
  <style
    global={true}
    jsx={true}
  >
    
              .nodes-epoch {
                background-color: #292526;
                color: #d5d4d8;
                font-size: 16px;
                font-weight: 500;
              }

              .nodes-epoch .nodes-epoch-content {
                margin: 15px 0;
              }

              .nodes-epoch .nodes-epoch-content .text-value {
                color: #37dbf4;
              }

              .nodes-epoch .nodes-epoch-content .text-value.persnt-remains {
                font-weight: 700;
              }

              .node-epoch-line-progress {
                padding-left: 0;
                padding-right: 0;
                height: 4px;
              }

              .node-epoch-circle-progress {
                margin-left: auto;
                width: 40px;
              }

              .node-epoch-circle-progress .circle-progress-label {
                color: #d5d4d8;
              }
            
  </style>
</div>
`;
